### PR TITLE
Email notifications

### PR DESCRIPTION
**Pull request by isocolsky**

### CODE REVIEW

File: test/integration/server.js [hunk 1]
```diff
@@ -3030,4 +3083,50 @@ describe('Wallet service', function() {
...
+          helpers.getAuthServer(copayer.id, function(server) {
+            server.savePreferences({
+              email: 'copayer' + (i++) + '@domain.com',
+            }, next);
+          });
+        }, done);
+      });
+    });
+
+    it('should notify copayers a new tx proposal has been created', function(done) {
```
Comment (matiu) [inline test/integration/server.js]: nice tests!

File: lib/server.js [hunk 2]
```diff
@@ -413,6 +437,43 @@ WalletService.prototype.joinWallet = function(opts, cb) {
...
+WalletService.prototype.savePreferences = function(opts, cb) {
+  var self = this;
+
+  opts = opts || {};
+
+  self._runLocked(cb, function(cb) {
+    var preferences = Model.Preferences.create({
+      walletId: self.walletId,
+      copayerId: self.copayerId,
+      email: opts.email,
```
Comment (matiu) [inline lib/server.js]: We should check `opts.email.length` < 254 , and `opts.email` is a valid email address.

File: lib/emailservice.js [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+EmailService.prototype.sendEmail = function(notification, cb) {
+  var self = this;
+
+  var emailType = EMAIL_TYPES[notification.type];
+  if (!emailType) return cb();
+
+  self._getRecipientsList(notification, emailType, function(err, recipientsList) {
+    if (_.isEmpty(recipientsList)) return cb();
+
+    async.waterfall([
```
Comment (matiu) [inline lib/emailservice.js]: nice!

**Comment by matiu** [comment]
ACK. Reviewed the CODE.

File: lib/server.js (hunk 2 above)
Comment (isocolsky) [inline lib/server.js]: I will be adding a validation for the max length. Validating the actual email address is a hard problem to solve (almost all regexes tag valid addresses as invalid, unicode email addresses, etc). Let's just trust the user to provide a valid email address.

Comment (matiu) [inline lib/server.js]: lets use a popular module like https://github.com/Sembiance/email-validator


On Mon, May 11, 2015 at 11:39 AM, Ivan Socolsky <notifications@github.com>
wrote:

> In lib/server.js
> <https://github.com/bitpay/bitcore-wallet-service/pull/207#discussion_r30043833>
> :
>
> > @@ -413,6 +437,43 @@ WalletService.prototype.joinWallet = function(opts, cb) {
> >  };
> >
> >  /**
> > + * Save copayer preferences for the current wallet/copayer pair.
> > + * @param {Object} opts
> > + * @param {string} opts.email - Email address for notifications.
> > + */
> > +WalletService.prototype.savePreferences = function(opts, cb) {
> > +  var self = this;
> > +
> > +  opts = opts || {};
> > +
> > +  self._runLocked(cb, function(cb) {
> > +    var preferences = Model.Preferences.create({
> > +      walletId: self.walletId,
> > +      copayerId: self.copayerId,
> > +      email: opts.email,
>
> I will be adding a validation for the max length. Validating the actual
> email address is a hard problem to solve (almost all regexes tag valid
> addresses as invalid, unicode email addresses, etc). Let's just trust the
> user to provide a valid email address.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitpay/bitcore-wallet-service/pull/207/files#r30043833>
> .
>



-- 
Matías Alejo Garcia
@ematiu
Roads? Where we're going, we don't need roads!

Comment (isocolsky) [inline lib/server.js]: Done!

**Comment by isocolsky** [comment]
Updated. Thank you very much reviewing and for the suggestions.

**Pull request merged**